File: src/main/java/de/thkoeln/inf/gpm/vgb/model/Address.java
package de.thkoeln.inf.gpm.vgb.model;

import model.AddressDao;
import util.DbUtil;

public class Address {
    private Long id;
    private String street;
    private String houseNumber;
    private Location location;

    public static Address createOrUpdate(Address address) {
        return DbUtil.INSTANCE.runInTransaction(() -> AddressDao.Companion.save(address));
    }

    public static Address findById(Long addressId) {
        return DbUtil.INSTANCE.runInTransaction(() -> AddressDao.Companion.get(addressId).toAddress());
    }

    public static void delete(Long addressId) {
        DbUtil.INSTANCE.runInTransaction(() -> AddressDao.Companion.delete(addressId));
    }

    public Address(Long id, String street, String houseNumber, Location location) {
        this.id = id;
        this.street = street;
        this.houseNumber = houseNumber;
        this.location = location;
    }

    public Address(String street, String houseNumber, Location location) {
        this.street = street;
        this.houseNumber = houseNumber;
        this.location = location;
    }

    public Long getId() {
        return id;
    }

    public String getStreet() {
        return street;
    }

    public void setStreet(String street) {
        this.street = street;
    }

    public String getHouseNumber() {
        return houseNumber;
    }

    public void setHouseNumber(String houseNumber) {
        this.houseNumber = houseNumber;
    }

    public Location getLocation() {
        return location;
    }

    public void setLocation(Location location) {
        this.location = location;
    }
}

File: src/main/java/de/thkoeln/inf/gpm/vgb/model/Address.kt
package de.thkoeln.inf.gpm.vgb.model

import org.jetbrains.exposed.dao.IntEntity
import org.jetbrains.exposed.dao.IntEntityClass
import org.jetbrains.exposed.dao.id.EntityID
import org.jetbrains.exposed.dao.id.IntIdTable
import org.jetbrains.exposed.sql.transactions.transaction

object Addresses : IntIdTable() {
    val street = text("Strasse")
    val houseNumber = text("Hausnummer")
    val location = reference("Ort", Locations)
}

class AddressDao(id: EntityID<Int>) : IntEntity(id) {
    private var street by Addresses.street
    private var houseNumber by Addresses.houseNumber
    private var location by LocationDao referencedOn Addresses.location

    companion object : IntEntityClass<AddressDao>(Addresses) {
        /**
         * Update or create [Address] in database
         */
        fun save(address: Address) : Address? = transaction {
            val location = LocationDao.save(address.location) ?: return@transaction null
            val locationDao = LocationDao.findById(location.id!!) ?: return@transaction null
            val newAddress = if (address.id == null) {
                new { update(address, locationDao) }
            } else {
                val old = findById(address.id)
                old?.update(address, locationDao)
                findById(address.id)
            }

            newAddress?.toAddress()
        }
    }

    private fun update(address: Address, locationDao: LocationDao) {
        this.street = address.street
        this.houseNumber = address.houseNumber
        this.location = locationDao
    }

    fun toAddress() = Address(id.value, street, houseNumber, location.toLocation())
}

class Address(
    val id: Int?,
    val street: String,
    val houseNumber: String,
    val location: Location
)